extensions/poetry/tests.ts

```diff
@@ -0,0 +1,152 @@
+  await copy('./fixtures',  phylum.fixturesDir)
+  // The `poetry_test` project should be pre-existing.
+  await phylum.run(['project', 'link', 'poetry_test'], phylum.fixturesDir)
+})
+
+afterAll(async () => {
+  await phylum.cleanup()
+})
+
+describe("Unit tests", async () => {
+  it("correctly parses installed packages", async () => {
+    const fixture = `
+      Updating dependencies
+      Resolving dependencies...
+         1: fact: fixture is 0.1.0
+         1: derived: fixture
+         1: fact: fixture depends on pandas (^1.4.3)
+         1: selecting fixture (0.1.0)
+         1: derived: pandas (>=1.4.3,<2.0.0)
+      PyPI: 1 packages found for pandas >=1.4.3,<2.0.0
+         1: fact: pandas (1.4.3) depends on python-dateutil (>=2.8.1)
+         1: fact: pandas (1.4.3) depends on pytz (>=2020.1)
+         1: fact: pandas (1.4.3) depends on numpy (>=1.21.0)
+         1: selecting pandas (1.4.3)
+         1: derived: numpy (>=1.21.0)
+         1: derived: pytz (>=2020.1)
+         1: derived: python-dateutil (>=2.8.1)
+      PyPI: No release information found for numpy-0.9.6, skipping
+      PyPI: No release information found for numpy-1.4.0, skipping
+      PyPI: 14 packages found for numpy >=1.21.0
+      PyPI: 6 packages found for pytz >=2020.1
+      PyPI: No release information found for python-dateutil-0.1, skipping
+      PyPI: No release information found for python-dateutil-2.0, skipping
+      PyPI: 2 packages found for python-dateutil >=2.8.1
+         1: fact: python-dateutil (2.8.2) depends on six (>=1.5)
+         1: selecting python-dateutil (2.8.2)
+         1: derived: six (>=1.5)
+      PyPI: 18 packages found for six >=1.5
+         1: selecting pytz (2022.1)
+         1: selecting six (1.16.0)
+         1: selecting numpy (1.23.1)
+         1: Version solving took 0.038 seconds.
+         1: Tried 1 solutions.
+    `
+
+    const parsed = Functions.parseDryRun(fixture)
+
+    assert(parsed.find(c => c.name === 'six' && c.version == '1.16.0'))
+    assert(parsed.find(c => c.name === 'numpy' && c.version == '1.23.1'))
+    assert(parsed.find(c => c.name === 'pytz' && c.version == '2022.1'))
+    assert(parsed.find(c => c.name === 'pandas' && c.version == '1.4.3'))
+  })
+})
+
+describe("Poetry extension", async () => {
+  // These tests may fail if the packages aren't processed on staging.
+
+  it("correctly handles the `--dry-run` argument", async () => {
+    let status = await phylum.runExt(['add', '--dry-run', 'pandas'], phylum.fixturesDir)
+    assert(status.code === 0)
+  })
+
+  it("correctly allows a valid package", async () => {
+    let status = await phylum.runExt(['add', 'pandas'], phylum.fixturesDir)
+    assert(status.code === 0)
+  })
+
+  // TODO tqdm is artificially marked as not passing; replace with an actual
+  // known-bad package.
+  it("correctly denies an invalid package", async () => {
+    let status = await phylum.runExt(['add', 'tqdm'], phylum.fixturesDir)
+    assert(status.code === 1)
+  })
```

Comment:
I'd like to explicitly point this out as something that should **NOT** be merged until it is resolved.

Reply:
Other tests are in a similar situation, where they are explicitly relying on the state of our service; those are not much better than this in my opinion.

Even once we have an explicitly failing Python package in staging or prod, we have no guarantee that that is going to be its permanent state.

The alternative would be bundling a mock API server but that's preposterous to even write down.

This test was explicitly requested in this same review and I agree it should be there.

Any suggestions?

Reply:
Which other test is relying on a temporarily manually set thing in our backend? If there are others, we should get rid of those too.

Honestly I don't particularly care about this test anyway, since it never gets executed, but really that's just one more reason why it shouldn't be kept in some garbage state indefinitely when nobody will ever touch it again.

Reply:
How about using `pyyaml==5.3.1` instead? That is my go-to known bad Python package.

Reply:
> Which other test is relying on a temporarily manually set thing in our backend? If there are others, we should get rid of those too.

I didn't say that. I said that both rely on the state of our service whose consistency we can't really guarantee. Besides, `tqdm` is the "bad dependency" in the mock API script I was using, not even on staging. Between a mock API script one can control and the wider service on staging, the latter is probably more reliable long term, provided we include that somewhere; I don't think we need to do that, as staging will be more than enough anyway for the purpose of these tests (see next paragraph) and reliability is not a concern.

> Honestly I don't particularly care about this test anyway, since it never gets executed, but really that's just one more reason why it shouldn't be kept in some garbage state indefinitely when nobody will ever touch it again.

These tests are meant for development workflow and I have been running them regularly. The fact that they aren't run in CI doesn't mean nobody runs them. If we ever want to put extensions test through CI, nobody is stopping us, though.

> How about using `pyyaml==5.3.1` instead? That is my go-to known bad Python package.

That's good to know going forward -- I did find one `yaml` package on Phylum but Poetry couldn't `add` that. Went with `cffi` in the meantime, which carries risk in Phylum due to the fact that it involves FFI.

Reply:
> These tests are meant for development workflow and I have been running them regularly. The fact that they aren't run in CI doesn't mean nobody runs them.

I strongly disagree with that. I certainly won't run them, you might but even that I'd be surprised about when you make changes to the CLI instead of the extension itself. And even then it's only one extension now, that would become unsustainable with more extensions.

So realistically nobody will ensure these extensions work consistently on every commit in master and it can potentially take a very long time before anyone tries to run them again.

> Besides, tqdm is the "bad dependency" in the mock API script I was using, not even on staging.

So that sounds even if I wanted to try running your tests, I couldn't? That sounds like even more of an issue. Having tests only a single developer knows how to execute sounds like a **TERRIBLE** idea.

Reply:
> I strongly disagree with that. I certainly won't run them, you might but even that I'd be surprised about when you make changes to the CLI instead of the extension itself.

That is not a good argument for throwing away test code that has proven useful for development and will prove useful during maintenance of the extension. To reiterate, I **do need** and **will need** those tests for continued maintenance, regardless of who runs them, and not having them would actively hamper my ability to work on the extension. If this PR gets blocked because of the presence of that test file, I will remove it and then simply proceed to waste valuable time by copying it outside of the repo and then back inside at every commit.

> And even then it's only one extension now, that would become unsustainable with more extensions.
>
> So realistically nobody will ensure these extensions work consistently on every commit in master and it can potentially take a very long time before anyone tries to run them again.

The unsustainability is easily addressed by setting up proper test execution in CI if and when we so desire (which, in my opinion, is not right here and right now, but that should be discussed elsewhere).

At some point we may even consider just moving the extensions directory to its own repository with its own CI pipeline, as extensions are not "code added to the CLI" but rather their own thing.

> So that sounds even if I wanted to try running your tests, I couldn't? That sounds like even more of an issue. Having tests only a single developer knows how to execute sounds like a **TERRIBLE** idea.

You can now, against staging, as I changed that package to a known failing package that's in staging.